src/mds/Server.cc

```diff
@@ -3534,10 +3544,57 @@ void Server::journal_allocated_inos(MDRequestRef& mdr, EMetaBlob *blob)
 		      mds->inotable->get_projected_version());
 }
 
-void Server::apply_allocated_inos(MDRequestRef& mdr, Session *session)
+void Server::try_delegate_inos(MDRequestRef& mdr, bool dirty) {
+  const cref_t<MClientRequest> &req = mdr->client_request;
+
+  if (req->get_op() == CEPH_MDS_OP_CREATE && mdr->tracei) {
+    if (mdr->did_early_reply) {
+      dout(20) << __func__ << ": inos probably delegated during early reply" << dendl;
+      return;
+    }
+
+    auto inode = mdr->tracei;
+    auto session = mdr->session;
+    dout(20) << __func__ << ": session=" << session << ", dirty=" << dirty
+	     << dendl;
+
+    // delay handing out preallocated inode range to client if the session
+    // needs to be persisted reflecting info.prealloc_inos to be saved
+    // before the client can start using them.
+    if (session->info.has_feature(CEPHFS_FEATURE_DELEG_INO)) {
+      if (!dirty) {
+	openc_response_t ocresp;
+	dout(10) << __func__ << ": adding created_ino and delegated_inos" << dendl;
+	ocresp.created_ino = inode->_get_inode()->ino;
+
+	if (delegate_inos_pct && !req->is_queued_for_replay()) {
+	  // Try to delegate some prealloc_inos to the client, if it's down to half the max
+	  unsigned frac = 100 / delegate_inos_pct;
+	  if (session->delegated_inos.size() < (unsigned)g_conf()->mds_client_prealloc_inos / frac / 2) {
+	    session->delegate_inos(g_conf()->mds_client_prealloc_inos / frac, ocresp.delegated_inos);
+	    dout(20) << __func__ << ": sending " << ocresp.delegated_inos.size()
+		     << " preallocated inodes" << dendl;
+	  }
+	}
+
+	encode(ocresp, mdr->reply_extra_bl);
+      } else {
+	dout(20) << __func__ << ": persisting sessionmap - not handing over preallocated inodes"
+		 << dendl;
+      }
+    } else if (req->get_connection()->has_feature(CEPH_FEATURE_REPLY_CREATE_INODE)) {
+      dout(10) <<  __func__ << ": adding ino to reply to indicate inode was created" << dendl;
+      // add the file created flag onto the reply if create_flags features is supported
+      encode(inode->ino(), mdr->reply_extra_bl);
```

Comment:
I think the `inode->ino()`  is exactly the same with `_inode->ino` or `inode->_get_inode()->ino`. If so could we make them consistent to make it more readable ?

Reply:
sure.